### ISSUE TITLE
VDiff false positive MismatchedRows when WEIGHT_STRING result is greater than max_allowed_packet

### ISSUE THREAD

**Issue by DeathBorn** [vitessio/vitess#7296]
#### Overview of the Issue

VDiff false positive MismatchedRows
When WEIGHT_STRING result is greater than max_allowed_packet, result is NULL.
If max_allowed_packet is high enough on  source or target, VDiff will compare NULL with BLOB - will report MismatchedRows.

#### Reproduction Steps

Steps to reproduce this issue, example:
Using local example with vitess version
```
curl -LJO https://github.com/planetscale/vitess-releases/releases/download/c02b7d5/vitess-8.0-dev-c02b7d5.tar.gz
```

```bash
./101_initial_cluster.sh
./201_customer_tablets.sh

# create table and data
mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000100/mysql.sock vt_commerce -e 'CREATE TABLE `mediumtext_test` (`id` int(11) NOT NULL AUTO_INCREMENT, `body` mediumtext COLLATE utf8mb4_unicode_ci, PRIMARY KEY (`id`) ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci;'
mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000100/mysql.sock vt_commerce -e "truncate table mediumtext_test";
mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000100/mysql.sock vt_commerce -e "insert into mediumtext_test (body) values (REPEAT('a',65535-1)) ";
mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000100/mysql.sock vt_commerce -e "insert into mediumtext_test (body) values (REPEAT('b',65535)) ";
mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000100/mysql.sock vt_commerce -e "insert into mediumtext_test (body) values (REPEAT('c',65535+1)) ";
mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000100/mysql.sock vt_commerce -e "insert into mediumtext_test (body) values (REPEAT('d',65535+2)) ";
mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000100/mysql.sock vt_commerce -e "insert into mediumtext_test (body) values (REPEAT('e',99999)) ";
mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000100/mysql.sock vt_commerce -e "select id, SUBSTRING(body, 1, 1), LENGTH(body) from mediumtext_test;"

# update vschema
cat > vschema_commerce_mediumtext.json<<'EOF'
{
    "tables": {
        "product": {},
        "customer": {},
        "corder": {},
        "mediumtext_test": {}
    }
}
EOF
vtctlclient ApplyVSchema -vschema_file vschema_commerce_mediumtext.json commerce
vtctlclient ReloadSchemaKeyspace commerce

# set low max_allowed_packet on target tablet mysql (2mb) 
mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000200/mysql.sock -e 'set global max_allowed_packet=2097152;';
mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000201/mysql.sock -e 'set global max_allowed_packet=2097152;';
mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000202/mysql.sock -e 'set global max_allowed_packet=2097152;';

# start workflow to move the table
vtctlclient -alsologtostderr=1 MoveTables  -workflow=commerce2customer commerce customer mediumtext_test


vtctlclient VReplicationExec zone1-000000100 "select * from _vt.copy_state"
+----------+------------+--------+
| vrepl_id | table_name | lastpk |
+----------+------------+--------+
+----------+------------+--------+


mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000100/mysql.sock vt_commerce -e "select id, SUBSTRING(body, 1, 1), LENGTH(body) from mediumtext_test;"
+----+-----------------------+--------------+
| id | SUBSTRING(body, 1, 1) | LENGTH(body) |
+----+-----------------------+--------------+
|  1 | a                     |        65534 |
|  2 | b                     |        65535 |
|  3 | c                     |        65536 |
|  4 | d                     |        65537 |
|  5 | e                     |        99999 |
+----+-----------------------+--------------+

mysql -S /home/centos/my-vitess-example/vtdataroot/vt_0000000200/mysql.sock vt_customer -e "select id, SUBSTRING(body, 1, 1), LENGTH(body) from mediumtext_test;"
+----+-----------------------+--------------+
| id | SUBSTRING(body, 1, 1) | LENGTH(body) |
+----+-----------------------+--------------+
|  1 | a                     |        65534 |
|  2 | b                     |        65535 |
|  3 | c                     |        65536 |
|  4 | d                     |        65537 |
|  5 | e                     |        99999 |
+----+-----------------------+--------------+

vtctlclient -alsologtostderr=1 VDiff customer.commerce2customer
I0114 16:16:53.001919   11402 main.go:64] I0114 16:16:53.000116 traffic_switcher.go:394] Migration ID for workflow commerce2customer: 6740760173349126720
E0114 16:16:53.109275   11402 main.go:64] E0114 16:16:53.090563 vdiff.go:894] [table=mediumtext_test] Different content 0 in same PK: [INT32(4) TEXT("d....d") BLOB("\x0em\x0em....\x0em")] != [INT32(4) TEXT("d....d") NULL]
E0114 16:16:53.125248   11402 main.go:64] E0114 16:16:53.106913 vdiff.go:894] [table=mediumtext_test] Different content 1 in same PK: [INT32(5) TEXT("e....e") BLOB("\x0e\x8bx0e.....e\x8b")] != [INT32(5) TEXT("e...e") NULL]
Summary for mediumtext_test: {ProcessedRows:5 MatchingRows:3 MismatchedRows:2 ExtraRowsSource:0 ExtraRowsTarget:0}
```
#### Binary version

```sh
vtgate --version
Version: c02b7d5 (Git branch 'master') built on Thu Sep 24 07:26:10 UTC 2020 by planetscale@ip-172-31-17-134 using go1.13.4 linux/amd64
```

#### Operating system and Environment details
 Centos

**Comment by tomas-didziokas** [comment]
Also, `VDiff` won't work as expected even when source and destination mysql instances has the same `max_allowed_packet`. For large enough values `WEIGHT_STRING` will return `NULL` from source and destination targets and `VDiff` will compare `NULL` vs `NULL` instead of comparing binary value representation.

**Comment by rohit-nayak-ps** [comment]
Closed via #7696
